src/internet_identity/src/active_anchor_stats.rs

```diff
@@ -0,0 +1,167 @@
+        // there was no activity for more than 24h since the end of the completed window so the
+        // last ongoing window is already outdated
+        // -> create an empty counter for the last completed window
+        if monthly_stats.start_timestamp + 31 * DAY_NS <= now {
+            // align empty completed 30-day window to the 24h collection interval
+            let offset = (now - monthly_stats.start_timestamp) % DAY_NS;
+            stats.completed.monthly_active_anchors =
+                Some(new_active_anchor_counter(now - 30 * DAY_NS - offset));
+        }
+    }
+
+    // Align new window to the currently running 24h interval
+    let start_timestamp = stats.ongoing.daily_active_anchors.start_timestamp;
+    // Note: requires daily stats to be processed before the monthly stats
+    assert!(now - start_timestamp < DAY_NS);
+
+    if let Some(monthly_stats) = stats.ongoing.monthly_active_anchors.last() {
+        if monthly_stats.start_timestamp + 1 * DAY_NS <= now {
+            // Start a new 30-day time window if the last one starts more than 24h in the past.
+            // This will result in at most 30 ongoing collection windows:
+            // * at most one window is added every 24h
+            // * windows that started more than 30 days ago are removed (see above)
+            // => within a 30-day period there can be at most 30 active windows offset by 24h each.
+            //
+            // Example using 3 day collection periods:
+            // Each character in <-> indicates a 24h window
+            // <->
+            //  <->
+            //   <->
+            //    <-> <- this window has no overlap with the first window anymore
+            //           -> for any time t there are at most 3 ongoing collection windows
+            stats
+                .ongoing
+                .monthly_active_anchors
+                .push(new_active_anchor_counter(start_timestamp));
+        }
+    } else {
+        // there was no activity for so long that there is no ongoing 30-day collection window anymore
+        // -> start collecting again (in sync with the daily active anchor stats)
+        // Note: requires daily stats to be processed before the monthly stats
+        stats
+            .ongoing
+            .monthly_active_anchors
+            .push(new_active_anchor_counter(start_timestamp));
+    }
+}
+
+fn increment_stats_on_activity(
+    stats: &mut ActiveAnchorStatistics,
+    previous_activity_timestamp: Option<Timestamp>,
+) {
+    // increase all counters and exit early if there is no previous activity
+    let Some(timestamp) = previous_activity_timestamp else {
+        stats.ongoing.daily_active_anchors.counter += 1;
+        stats.ongoing.monthly_active_anchors.iter_mut().for_each(|monthly_stats| monthly_stats.counter +=1 );
+        return;
+    };
+
+    if stats.ongoing.daily_active_anchors.start_timestamp > timestamp {
```

Comment:
nit: maybe add a comment explaining how this results in counting each anchor only once per period, as this seems not quite obvious.